### PR TITLE
feat(portal): map a subscription form to a set of APIs

### PR DESCRIPTION
**Pull request by thomasballivet**
**Issue**

https://gravitee.atlassian.net/browse/PORTAL-168

**Description**

A form can be mapped to a set of APIs: an API is mapped to at most one form, and consumers
subscribing to it see that form.

- **Explicit mapping, no runtime fallback** — at upgrade, the former single form of the environment
  (now the "Global Default Form") is mapped to every existing API, so they keep the form they were
  shown. From then on every form, the Global Default Form included, is mapped manually: an API
  created afterwards has no form until it is mapped to one.

- **One resolver** — `SubscriptionFormResolutionDomainService` decides which form applies; the
  portal endpoint and subscription validation both go through it, so the consumer and the server
  see the same form.
- **A hidden form yields no form at all** for its APIs. The Console says so when you hide one.
- **Console, per the design** — a "Mapped APIs" section under the editor: chips for the mapped APIs,
  and a searchable, paginated list of every API of the environment (version, "Mapped" badge) to map
  or unmap with a checkbox. An API already mapped to another form is disabled, with a tooltip naming
  that form.
- **No `gio-table-wrapper` in that section** — its header bar (search + full paginator) overflowed
  the panel (674 px of content in 643 px) and added a second horizontal scrollbar. The section lays
  out its own search field above the table and a single paginator below it.
- **Toggle preview**, per the design — a button in the editor header hides the GMD preview so the
  editor takes the full width. The preview is hidden, not removed: the fields it renders feed the
  validation panel, and removing them reported every form as valid.

`apiIds` is required on `PUT /subscription-forms/{id}`: a PUT replaces the mapping like it replaces
the name and the definition.

**Additional context**

Schema change: `subscription_form_apis` (Liquibase `v4_13_0/22`, unique on `api_id`) and `apiIds`
on Mongo with a unique index. Liquibase `v4_13_0/23` and a Mongo upgrader map the Global Default
Form to the existing APIs. The JDBC repository writes a form and its mappings in one
transaction; a collision answers 409.

Two commits touch the shared `gravitee-markdown` library. `gmd-form-editor` gains a `showPreview`
input, `true` by default, so existing usages are unchanged. `gmd-monaco-editor` gets a fix for a bug
already on master: after saving a document, opening another one and coming back, the editor showed
empty while the preview kept the content. The navigation page editor uses the same component and
benefits from it too.

https://github.com/user-attachments/assets/b72040ea-a112-456e-ac07-be88314dbbdc

### CODE REVIEW

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/subscriptionforms/ApiIdsIndexUpgrader.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @author GraviteeSource Team
+ */
+@Component("SubscriptionFormsApiIdsIndexUpgrader")
+public class ApiIdsIndexUpgrader extends IndexUpgrader {
+
+    @Override
+    protected Index buildIndex() {
+        return Index.builder()
+            .collection("subscription_forms")
+            .name("ai1_unique")
```
[the remark this answers is not in the document]
Comment (dev-marek) [reply gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/subscriptionforms/ApiIdsIndexUpgrader.java:39]: @AikidoSec ignore: This is a project convention to use cryptic index names

[the remark this answers is not in the document]
Comment (dev-marek) [reply gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/subscriptionforms/ApiIdsIndexUpgrader.java:39]: @thomasballivet is `_unique` a part of our convention?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/subscription_form/domain_service/SubscriptionFormResolutionDomainService.java [hunk 2]
```diff
@@ -17,19 +17,23 @@
...
  */
 @DomainService
+@RequiredArgsConstructor
 public class SubscriptionFormResolutionDomainService {
 
+    private final SubscriptionFormQueryService subscriptionFormQueryService;
+
     public Optional<SubscriptionForm> resolveForApi(String environmentId, String apiId) {
-        return Optional.empty();
+        return subscriptionFormQueryService.findByApiId(environmentId, apiId).filter(SubscriptionForm::isEnabled);
```
Comment (dev-marek) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/subscription_form/domain_service/SubscriptionFormResolutionDomainService.java]: **Blocker — goal-fit.** This only resolves a form the API is *explicitly* mapped to via `subscription_form_apis`. The PR description states as delivered behavior that "every other API keeps the environment default" and that the default form "already covers every API without a dedicated form" — but there's no fallback here to a default/catalog form, and no `isDefault`/similar flag exists anywhere in the model, OpenAPI spec, or Console to even represent that concept at runtime.

This method's own tests confirm the gap is intentional in the current implementation: `should_return_nothing_when_the_api_has_no_dedicated_form` and `should_return_nothing_when_the_environment_has_no_form` both assert `isEmpty()`. The "Global Default Form" is realized only as a one-time migration (v4_13_0/23 + the Mongo upgrader) that backfills mappings for *pre-existing* APIs at upgrade time — it is not kept in sync afterward.

Consequence: any API created in an environment *after* this feature ships gets no subscription form at all, silently regressing the pre-catalog behavior where one form covered every API of the environment, until someone manually maps it in the Console.

Please either implement the described fallback (and the Console's "default form shows no mapping section" UI), or correct the PR description/changelog so this known gap is explicit and confirmed with stakeholders.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-model/src/main/resources/openapi/openapi-environments.yaml [hunk 3]
```diff
@@ -3367,9 +3379,18 @@ components:
...
+          description: |-
+            Identifiers of the APIs this form is dedicated to, replacing the current ones like the name and the
+            definition. Required: an absent or empty list unmaps every API. Each must belong to the environment
+            and to no other form.
+          items:
+            type: string
       required:
         - name
         - gmdContent
+        - apiIds
```
Comment (dev-marek) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-model/src/main/resources/openapi/openapi-environments.yaml:3396]: **Improvement.** `apiIds` is marked `required` here on `UpdateSubscriptionForm`, matching the PR description's "apiIds is required on PUT". But every existing PUT test in `SubscriptionFormResourceTest.java` always supplies `apiIds` (even if just `List.of()`) — none asserts that omitting it actually gets rejected (400). This is the one behavior the PR description explicitly calls out for PUT, and it currently has no test that would fail if the `required` annotation were accidentally dropped from the spec or generator config.

Suggest adding a resource test that PUTs with `apiIds` unset/null and asserts a 400.

File: gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    </mat-chip-set>
+  }
+
+  <mat-form-field class="mapped-apis__search" appearance="outline" subscriptSizing="dynamic">
+    <mat-icon matIconPrefix svgIcon="gio:search"></mat-icon>
+    <mat-label>Search APIs...</mat-label>
+    <input matInput type="search" [formControl]="searchControl" data-testid="api-search-input" />
+  </mat-form-field>
+
+  <table mat-table [dataSource]="apis()" aria-label="All available APIs">
```
Comment (dev-marek) [inline gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html:47]: **Nit.** This reimplements its own search field + `mat-table` + single `mat-paginator` instead of the shared `gio-table-wrapper`, which this module's `AGENTS.md` calls out as the established shared component for exactly this search+paginate+table pattern. The PR description explains this was deliberate (the shared component's own header-bar paginator overflowed the 643px panel), but that rationale only lives in the PR description, not in the code — a future maintainer reading just this file has no signal it's an intentional, justified deviation rather than an oversight.

Suggest a one-line comment here (or in the `.ts`) stating the overflow reason, mirroring how the rest of the file already documents non-obvious choices.

File: gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        <span
+          [attr.data-testid]="'api-mapped-elsewhere-' + api.id"
+          [matTooltip]="'Mapped to ' + mappedTo(api)"
+          [matTooltipDisabled]="!mappedTo(api)"
+        >
+          <mat-checkbox
+            [checked]="isSelected(api)"
+            [disabled]="!canUpdate() || !!mappedTo(api)"
+            [attr.data-testid]="'api-checkbox-' + api.id"
+            [aria-label]="'Map ' + api.name"
```
Comment (dev-marek) [inline gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html]: **Nit.** `[aria-label]="'Map ' + api.name"` is static regardless of `isSelected(api)`, so a screen-reader user gets no indication that clicking an already-checked box means "unmap" rather than "map". Minor since the checkbox's own `checked` state is announced separately, but easy to improve:

```html
[aria-label]="(isSelected(api) ? 'Unmap ' : 'Map ') + api.name"
```

**Review by dev-marek — CHANGES_REQUESTED** [review]
**Verdict: Request changes** — one confirmed silent-data-loss bug and one significant gap between the PR's stated design and what's actually implemented. Everything else (JDBC repo/migrations, Mongo upgraders, 409 exception mapping, the gravitee-markdown library changes) checked out clean and well-tested.

### Blocker 1 — `apiIds` silently dropped on form creation (not inline-commentable: the buggy lines aren't part of this diff)
`gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/crud_service/subscription_form/SubscriptionFormCrudServiceImpl.java:46-54`

```java
public SubscriptionForm create(SubscriptionForm subscriptionForm) {
    SubscriptionForm toCreate = subscriptionForm.getId() == null
        ? SubscriptionForm.builder()
            .id(SubscriptionFormId.random())
            .environmentId(subscriptionForm.getEnvironmentId())
            .name(subscriptionForm.getName())
            .gmdContent(subscriptionForm.getGmdContent())
            .enabled(subscriptionForm.isEnabled())
            .validationConstraints(subscriptionForm.getValidationConstraints())
            .build()
        : subscriptionForm;
```

`toCreate` never calls `.apiIds(...)`. Since `apiIds` is `@Builder.Default = List.of()` on the core `SubscriptionForm` model, any APIs a user selects while creating a *new* form (the Console's "new" editor state shows the mapping section and `createForm()` sends `apiIds`) are silently discarded before reaching the DB — the form is created with zero mapped APIs.

This is masked in tests: `SubscriptionFormCrudServiceInMemory` (the fake used by `CreateSubscriptionFormUseCaseTest`) *was* correctly updated to preserve `apiIds`, so `should_create_a_form_dedicated_to_apis` passes against the fake while the real path is broken. `SubscriptionFormCrudServiceImplTest`, which exercises the real class, wasn't touched by this PR and has no `apiIds` coverage at all.

**Fix:** add `.apiIds(subscriptionForm.getApiIds())` to the builder, and add a test on `SubscriptionFormCrudServiceImplTest` asserting it round-trips.

### What's solid
JDBC repo + migrations (transactional write of form + mappings, correct unique constraint on `api_id`), Mongo upgraders (correct ordering, correctly scoped backfill, well-tested), 409 handling via the existing generic `ConflictDomainExceptionMapper`, the `gmd-monaco-editor` reopen-after-save fix (minimal, root-caused, provably tested), and `gmd-form-editor`'s `showPreview`/`[hidden]` behavior (verified by a test that would catch a regression to `*ngIf`). Console component tests assert real behavior (disabled-state + tooltip, exact PUT payloads, unsaved-changes tracking) rather than just rendering.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/environment/SubscriptionFormResource.java [hunk 6]
```diff
@@ -88,7 +88,8 @@ public Response updateSubscriptionForm(@Valid @NotNull final UpdateSubscriptionF
                 GraviteeContext.getCurrentEnvironment(),
                 subscriptionFormId(),
                 request.getName(),
-                request.getGmdContent()
+                request.getGmdContent(),
+                request.getApiIds()
```
[the remark this answers is not in the document]
Comment (thomasballivet) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/environment/SubscriptionFormResource.java:92]: @AikidoSec ignore: deliberate: a PUT replaces the whole form, mapping included, and the Console is the only consumer of this endpoint, shipped in the same release.

File: gravitee-apim-console-webui/src/portal/subscription-form/subscription-form.component.ts [hunk 7]
```diff
@@ -179,7 +205,12 @@ export class SubscriptionFormComponent implements HasUnsavedChanges {
     if (this.selectedFormId() === null) return false;
     const currentName = (this.nameValue() ?? '').trim();
     const currentContent = normalizeContent(this.contentValue());
-    return currentName !== this.initialName().trim() || currentContent !== normalizeContent(this.initialContent());
+    const currentApiIds = this.selectedApis().map(api => api.id);
+    return (
+      currentName !== this.initialName().trim() ||
+      currentContent !== normalizeContent(this.initialContent()) ||
+      !sameIds(currentApiIds, this.initialApiIds())
+    );
```
[the remark this answers is not in the document]
Comment (thomasballivet) [reply gravitee-apim-console-webui/src/portal/subscription-form/subscription-form.component.ts:213]: @AikidoSec ignore: style preference, the single boolean expression reads fine here.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/crud_service/subscription_form/SubscriptionFormCrudServiceImpl.java
Comment (dev-marek) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/crud_service/subscription_form/SubscriptionFormCrudServiceImpl.java:47]: please remember about api ids

Comment (thomasballivet) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/crud_service/subscription_form/SubscriptionFormCrudServiceImpl.java:47]: Fixed in 5cfba076e4: `create` now keeps the `apiIds` when it generates the id, covered in `SubscriptionFormCrudServiceImplTest`.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-model/src/main/resources/openapi/openapi-environments.yaml (hunk 3 above)
Comment (thomasballivet) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-model/src/main/resources/openapi/openapi-environments.yaml:3396]: Good catch, it wasn't enforced: the generated model defaulted an absent `apiIds` to an empty list, so a PUT without it silently unmapped every API instead of failing. The property is now `nullable` in the spec, which keeps it null so the `@NotNull` generated by `required` answers 400, and a test covers it (7c303b01a3).

File: gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html (hunk 4 above)
Comment (thomasballivet) [reply gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html:47]: Added a comment in the template explaining it (7e88fa9eaa).

File: gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html (hunk 5 above)
Comment (thomasballivet) [reply gravitee-apim-console-webui/src/portal/subscription-form/subscription-form-apis/subscription-form-apis.component.html]: Done, with a test (7e88fa9eaa).

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/subscriptionforms/ApiIdsIndexUpgrader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasballivet) [reply gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/subscriptionforms/ApiIdsIndexUpgrader.java:39]: Not consistently: some unique indexes carry no suffix (`e1c1`, `k1`, `iu1iri1irt1`), others do (`ei1mci1_unique` on applications, the users one, `ei1nn1_unique`). I kept it to match the other unique index of this collection.

**Comment by thomasballivet** [comment]
Thanks @dev-marek, review addressed:

- **Blocker**: `SubscriptionFormCrudServiceImpl.create` now keeps the `apiIds` when it generates the id, with a test on the real class (5cfba076e4).
- Your 400 test on `PUT` also revealed that an absent `apiIds` silently unmapped every API: it is now rejected (7c303b01a3).

Nits done, branch rebased on master.

**Review by dev-marek — APPROVED** [review]
